### PR TITLE
add podgroup controller

### LINKED ISSUE #134: Add PodGroupController to creat shadow PodGroup
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**Description**:

Currently, kube-batch create shadow PodGroup by pod's `OwnerReference` for upstream objects, e.g. Deployment. It make Queue related feature harder, e.g. Queue's status, it's better to have such a controller to create `PodGroup` for upstream objects.

### PR DESCRIPTION
**Pull request by wangyuqing4**
Fixes #134

if normal pod/job/...... use volcano scheduler, podgroup controller watch pod, can create podgroup.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
Can u add desc about the relation with #165 and #370 ?  I donot know which one to review

**Comment by wangyuqing4** [comment]
#165 is huge, so split smaller one, #401 is part 1, #370 is part 3.

image

**Comment by hzxuzhonghu** [comment]
ok

File: pkg/controllers/podgroup/pg_controller.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	cc.podInformer = cc.sharedInformers.Core().V1().Pods()
+	cc.podLister = cc.podInformer.Lister()
+	cc.podSynced = cc.podInformer.Informer().HasSynced
+	cc.podInformer.Informer().AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch obj.(type) {
+				case *v1.Pod:
+					pod := obj.(*v1.Pod)
+					if pod.Spec.SchedulerName == "volcano" &&
```
Comment (k82cn) [inline pkg/controllers/podgroup/pg_controller.go]: We should not hard code scheduler name here.

Comment (wangyuqing4) [reply pkg/controllers/podgroup/pg_controller.go]: done

File: pkg/controllers/podgroup/pg_controller.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	schedulerName string,
+) *Controller {
+	cc := &Controller{
+		kubeClients: kubeClient,
+		kbClients:   kbClient,
+
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	cc.sharedInformers = informers.NewSharedInformerFactory(cc.kubeClients, 0)
```
Comment (hzxuzhonghu) [inline pkg/controllers/podgroup/pg_controller.go]: One point: we should prevent duplicate informers in different controllers. It is very costly

Comment (wangyuqing4) [reply pkg/controllers/podgroup/pg_controller.go]: ok，follow-up pr will optimize the part，gc/queue/job/pg controller will delete same code.

Comment (hzxuzhonghu) [reply pkg/controllers/podgroup/pg_controller.go]: agreed, please file an issue to track

File: pkg/controllers/podgroup/pg_controller.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	cc.pgSynced = cc.pgInformer.Informer().HasSynced
+
+	return cc
+}
+
+// Run start NewPodgroupController
+func (cc *Controller) Run(stopCh <-chan struct{}) {
+	go cc.sharedInformers.Start(stopCh)
+	go cc.podInformer.Informer().Run(stopCh)
+	go cc.pgInformer.Informer().Run(stopCh)
```
Comment (hzxuzhonghu) [inline pkg/controllers/podgroup/pg_controller.go]: We donot need Run informers separately as sharedInformers Start will Run the informers created from informer factory.

Comment (hzxuzhonghu) [reply pkg/controllers/podgroup/pg_controller.go]: Same happens to other controllers

File: pkg/controllers/podgroup/pg_controller_handler.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+func (cc *Controller) addPod(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		glog.Errorf("Failed to convert %v to v1.Pod", obj)
+		return
+	}
+
+	req := podRequest{
+		pod: pod,
```
Comment (hzxuzhonghu) [inline pkg/controllers/podgroup/pg_controller_handler.go]: Actually, you just need the pod ns/name

File: pkg/controllers/podgroup/pg_controller.go (hunk 3 above)
Comment (hzxuzhonghu) [inline pkg/controllers/podgroup/pg_controller.go]: `cc.sharedInformers.Star` will run all its informers.

File: pkg/controllers/podgroup/pg_controller_handler.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	vkbatchv1 "volcano.sh/volcano/pkg/apis/batch/v1alpha1"
+	scheduling "volcano.sh/volcano/pkg/apis/scheduling/v1alpha2"
+)
+
+type podRequest struct {
+	pod *v1.Pod
```
Comment (hzxuzhonghu) [inline pkg/controllers/podgroup/pg_controller_handler.go]: only need pod ns/name

File: pkg/controllers/podgroup/pg_controller.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	cc.pgLister = cc.pgInformer.Lister()
+	cc.pgSynced = cc.pgInformer.Informer().HasSynced
+
+	return cc
+}
+
+// Run start NewPodgroupController
+func (cc *Controller) Run(stopCh <-chan struct{}) {
+	go cc.sharedInformers.Start(stopCh)
+	go cc.podInformer.Informer().Run(stopCh)
```
Comment (k82cn) [inline pkg/controllers/podgroup/pg_controller.go]: if start sharedInformer, it's not necessary to start pod informer again.

Comment (hzxuzhonghu) [reply pkg/controllers/podgroup/pg_controller.go]: same thing happens to other controllers

File: pkg/controllers/podgroup/pg_controller_handler.go (hunk 5 above)
Comment (k82cn) [reply pkg/controllers/podgroup/pg_controller_handler.go]: We can not use pointer here: the same pod may have different object in cache.

File: pkg/controllers/podgroup/pg_controller_handler.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+func (cc *Controller) updatePodAnnotations(pod *v1.Pod, pgName string) error {
+	if pod.Annotations == nil {
+		pod.Annotations = make(map[string]string)
+	}
+	if pod.Annotations[scheduling.GroupNameAnnotationKey] == "" {
+		pod.Annotations[scheduling.GroupNameAnnotationKey] = pgName
+	} else {
+		return nil
```
Comment (k82cn) [inline pkg/controllers/podgroup/pg_controller_handler.go]: log an error message if `pod.Annotations[scheduling.GroupNameAnnotationKey] != pgName`.

File: pkg/controllers/podgroup/pg_controller_handler.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+			glog.Errorf("Failed to get normal PodGroup for Pod <%s/%s>: %v",
+				pod.Namespace, pod.Name, err)
+			return err
+		}
+
+		pg := &scheduling.PodGroup{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:       pod.Namespace,
+				Name:            pgName,
+				OwnerReferences: pod.OwnerReferences,
```
Comment (k82cn) [inline pkg/controllers/podgroup/pg_controller_handler.go]: if pod's ownerReferences is empty, set PodGroup's ower references to this Pod for GC.

File: pkg/controllers/podgroup/pg_controller_handler.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+
+	return cc.updatePodAnnotations(pod, pgName)
+}
+
+func generatePodgroupName(pod *v1.Pod) string {
+	pgName := vkbatchv1.PodgroupNamePrefix
+	if len(pod.OwnerReferences) != 0 {
+		pgName += string(pod.OwnerReferences[0].UID)
```
Comment (k82cn) [inline pkg/controllers/podgroup/pg_controller_handler.go]: we should find the controlled owner

File: pkg/controllers/podgroup/pg_controller_handler.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	pgName += string(pod.UID)
+
+	return pgName
+}
+
+func newPGOwnerReferences(pod *v1.Pod) []metav1.OwnerReference {
+	if len(pod.OwnerReferences) != 0 {
+		return pod.OwnerReferences
```
Comment (k82cn) [inline pkg/controllers/podgroup/pg_controller_handler.go]: there maybe not controlled ownerference there.

**Pull request merged**